# Lecture 2 — Class Exercise
# Bar Charts: World Happiness Report 2023

---

> **Your task:** Create **2 polished bar charts** using the World Happiness Report dataset.  
> **Push to:** `week02/lecture02_exercise.ipynb` in **your own GitHub repo** before the end of class.

---

### Rules to keep in mind 💡
1. Every bar chart **must have a zero baseline** — no exceptions
2. Every chart **must have an insight title**, not a topic title
3. Aim for **professional quality** — clean background, readable font, no clutter
4. Horizontal bars for long category names

---


## Setup — Run this cell first


In [12]:
import pandas as pd
import numpy as np

# World Happiness Report 2023 — representative data
# Source: https://www.kaggle.com/datasets/ajaypalsinghlo/world-happiness-report-2023

df = pd.read_csv('/content/world_happiness_2023.csv')

df.columns = ['Country','Region','Happiness_Score','GDP','Social_Support',
              'Life_Expectancy','Freedom','Generosity','Corruption']


print(f"Dataset: {len(df)} countries, {len(df.columns)} columns")
print(df.head())


Dataset: 63 countries, 9 columns
       Country                        Region  Happiness_Score     GDP  \
0      Finland                Western Europe            7.804  10.775   
1      Denmark                Western Europe            7.586  10.933   
2      Iceland                Western Europe            7.525  10.878   
3       Israel  Middle East and North Africa            7.473  10.527   
4  Netherlands                Western Europe            7.464  11.015   

   Social_Support  Life_Expectancy  Freedom  Generosity  Corruption  
0           0.954             71.9    0.949       0.142       0.179  
1           0.954             72.7    0.931       0.168       0.234  
2           0.983             72.5    0.961       0.260       0.150  
3           0.916             72.4    0.903       0.149       0.826  
4           0.939             72.4    0.879       0.240       0.296  


In [13]:
import plotly.express as px
import plotly.graph_objects as go

# Explore the dataset before you start
print("Regions in dataset:")
print(df['Region'].value_counts())
print("\nScore range:", df['Happiness_Score'].min(), "–", df['Happiness_Score'].max())
print("\nBottom 10 countries:")
print(df.nsmallest(10, 'Happiness_Score')[['Country','Region','Happiness_Score']])


Regions in dataset:
Region
Western Europe                  15
Latin America and Caribbean     13
Central and Eastern Europe       7
Sub-Saharan Africa               7
Middle East and North Africa     6
North America and ANZ            4
Southeast Asia                   4
South Asia                       4
East Asia                        3
Name: count, dtype: int64

Score range: 1.859 – 7.804

Bottom 10 countries:
        Country                        Region  Happiness_Score
60  Afghanistan                    South Asia            1.859
61      Lebanon  Middle East and North Africa            2.392
62     Zimbabwe            Sub-Saharan Africa            2.995
52     Ethiopia            Sub-Saharan Africa            3.564
53     Tanzania            Sub-Saharan Africa            3.698
48   Bangladesh                    South Asia            3.892
47        India                    South Asia            4.036
50        Kenya            Sub-Saharan Africa            4.112
54       Uganda

---
## Task 1 — Regional Comparison Bar Chart

**What to build:** A horizontal bar chart showing the **average happiness score by region**, sorted from highest to lowest.

**Requirements:**
- Horizontal orientation (category names are long)
- Sorted by score, descending (so the happiest region is at the top)
- Zero baseline on x-axis
- At least one design choice that goes beyond the Plotly default (colour, annotation, labels, etc.)
- An insight title that answers: *which region stands out and why does it matter?*

**Hint:** Use `df.groupby('Region')['Happiness_Score'].mean()` to compute the averages.


In [14]:
# Task 1: Average happiness by region
# -----------------------------------

# Calculate the average happiness score for each region
region_avg = (
    df.groupby('Region')['Happiness_Score']
      .mean()
      .reset_index()
      .sort_values('Happiness_Score', ascending=True)
)

print(region_avg)

# Create horizontal bar chart
fig = px.bar(
    region_avg,
    x='Happiness_Score',
    y='Region',
    orientation='h',
    text='Happiness_Score',
    title='Western Europe has the highest average happiness score',
    labels={
        'Happiness_Score': 'Average Happiness Score',
        'Region': ''
    }
)

# Add values to the bars
fig.update_traces(
    marker_color='cornflowerblue',
    texttemplate='%{x:.2f}',
    textposition='outside'
)

# Start x-axis from zero
fig.update_xaxes(
    range=[0, 8],
    showgrid=True,
    gridcolor='lightgray'
)

fig.update_layout(
    template='simple_white',
    height=550,
    title_font_size=20,
    margin=dict(l=20, r=80, t=80, b=40)
)

fig.show()




                         Region  Happiness_Score
5                    South Asia         3.618250
7            Sub-Saharan Africa         4.064714
3  Middle East and North Africa         4.943333
6                Southeast Asia         5.695250
2   Latin America and Caribbean         5.699000
1                     East Asia         5.966000
0    Central and Eastern Europe         6.338143
4         North America and ANZ         7.018250
8                Western Europe         7.085533


### Task 1 — Insight

Western Europe has the highest average happiness score at around 7.09. South Asia has the lowest score at about 3.62. The difference between the two regions is quite large.

---
## Task 2 — Bottom vs. Top: A Contrast Story

**What to build:** A bar chart that highlights the **gap between the happiest and least happy countries**, focusing on a specific insight.

**Requirements:**
- Show the **top 8 AND bottom 8 countries** together (16 bars total)
- Use **colour** to distinguish the two groups (not Plotly's default rainbow)
- Add a **visual separator or annotation** that emphasises the gap
- Insight title that tells the story of the gap

**Hint:** Use `pd.concat([df.nlargest(8,'Happiness_Score'), df.nsmallest(8,'Happiness_Score')])` to get both groups.

**Stretch goal:** Add a vertical reference line showing the global average.


In [15]:
# Task 2: Comparing the happiest and least happy countries
# ----------------------------------------------------------

# Select the 8 happiest and 8 least happy countries
top8 = df.nlargest(8, 'Happiness_Score').copy()
top8['Group'] = 'Happiest 8'

bottom8 = df.nsmallest(8, 'Happiness_Score').copy()
bottom8['Group'] = 'Least Happy 8'

# Combine both groups
comparison = pd.concat([bottom8, top8])
comparison = comparison.sort_values('Happiness_Score')

# Calculate global average
global_average = df['Happiness_Score'].mean()

print(f"Global average happiness score: {global_average:.2f}")

# Create the chart
fig = px.bar(
    comparison,
    x='Happiness_Score',
    y='Country',
    orientation='h',
    color='Group',
    text='Happiness_Score',
    color_discrete_map={
        'Happiest 8': 'seagreen',
        'Least Happy 8': 'salmon'
    },
    title='The gap between the happiest and least happy countries is clear',
    labels={
        'Happiness_Score': 'Happiness Score',
        'Country': '',
        'Group': ''
    }
)

fig.update_traces(
    texttemplate='%{x:.2f}',
    textposition='outside'
)

# Zero baseline
fig.update_xaxes(
    range=[0, 8.5],
    showgrid=True,
    gridcolor='lightgray'
)

# Global average
fig.add_vline(
    x=global_average,
    line_dash='dash',
    line_width=2,
    annotation_text=f'Global average = {global_average:.2f}',
    annotation_position='top'
)

fig.update_layout(
    template='simple_white',
    height=700,
    title_font_size=20,
    margin=dict(l=20, r=100, t=90, b=40)
)

fig.show()

Global average happiness score: 5.81


### Task 2 — Insight

The difference between the two groups is easy to see. The happiest 8 countries are all above the global average of 5.81, while the least happy 8 are below it. This shows a clear gap in happiness scores between the two groups.